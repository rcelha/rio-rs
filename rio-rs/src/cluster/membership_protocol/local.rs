//! In-memory ClusterProvider for testing.

use async_trait::async_trait;
use std::time::Duration;

use super::ClusterProvider;
use crate::{cluster::storage::MembersStorage, errors::ClusterProviderServeError};

/// Local server compatible with the ClusterProvider API
///
/// This is only for tests, and it is doesn't offer real
/// cluster capabilities
#[derive(Clone)]
pub struct LocalClusterProvider<T> {
    pub members_storage: T,
}

#[async_trait]
impl<T> ClusterProvider<T> for LocalClusterProvider<T>
where
    T: MembersStorage,
{
    fn members_storage(&self) -> &T {
        &self.members_storage
    }

    async fn serve(&self, _address: &str) -> Result<(), ClusterProviderServeError> {
        loop {
            tokio::time::sleep(Duration::from_secs(1)).await;
        }
    }
}
