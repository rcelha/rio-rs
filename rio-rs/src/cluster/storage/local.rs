//! In-Memory implementation of [MembersStorage]
use std::sync::Arc;

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use tokio::sync::RwLock;

use super::{Member, MembersStorage, MembershipResult, MembershipUnitResult};

type ArcMembers = Arc<RwLock<Vec<Member>>>;
type ArcFailures = Arc<RwLock<Vec<(String, String, DateTime<Utc>)>>>;

#[derive(Clone, Default)]
pub struct LocalStorage {
    members: ArcMembers,
    failures: ArcFailures,
}

#[async_trait]
impl MembersStorage for LocalStorage {
    async fn push(&self, member: Member) -> MembershipUnitResult {
        self.members.write().await.push(member);
        Ok(())
    }

    async fn remove(&self, ip: &str, port: &str) -> MembershipUnitResult {
        let mut guard = self.members.write().await;
        guard.retain(|x| x.ip() != ip || x.port() != port);
        Ok(())
    }

    async fn set_is_active(&self, ip: &str, port: &str, is_active: bool) -> MembershipUnitResult {
        let last_seen = Utc::now();
        let mut guard = self.members.write().await;
        for i in guard.iter_mut() {
            if i.ip() == ip && i.port() == port {
                i.set_active(is_active);
                i.last_seen = last_seen.clone();
            }
        }
        Ok(())
    }

    async fn notify_failure(&self, ip: &str, port: &str) -> MembershipUnitResult {
        let now = Utc::now();
        let mut guard = self.failures.write().await;
        guard.push((ip.to_string(), port.to_string(), now));
        Ok(())
    }

    async fn member_failures(&self, ip: &str, port: &str) -> MembershipResult<Vec<DateTime<Utc>>> {
        let guard = self.failures.read().await;
        let items = guard
            .iter()
            .filter(|(ip_, port_, ..)| ip_ == ip && port_ == port)
            .cloned()
            .map(|x| x.2)
            .collect();
        Ok(items)
    }

    async fn members(&self) -> MembershipResult<Vec<Member>> {
        Ok(self.members.read().await.clone())
    }
}

#[cfg(test)]
mod test {
    use super::*;

    async fn storage() -> impl MembersStorage {
        let storage = LocalStorage::default();
        for (ip, port) in [
            ("0.0.0.0", "5000"),
            ("0.0.0.0", "5001"),
            ("0.0.0.0", "5002"),
            ("0.0.0.0", "5003"),
            ("0.0.0.0", "5004"),
            ("0.0.0.0", "5005"),
        ] {
            storage
                .push(Member::new(ip.to_string(), port.to_string()))
                .await
                .unwrap();
        }
        storage
    }

    #[tokio::test]
    async fn local_storage_clone_keep_members() {
        let storage = storage().await;
        let second_storage = storage.clone();

        let members = storage.members().await.unwrap();
        assert_eq!(members.len(), 6);
        let members = second_storage.members().await.unwrap();
        assert_eq!(members.len(), 6);

        second_storage.remove("0.0.0.0", "5005").await.unwrap();

        let members = storage.members().await.unwrap();
        assert_eq!(members.len(), 5);
        let members = second_storage.members().await.unwrap();
        assert_eq!(members.len(), 5);
    }
}
