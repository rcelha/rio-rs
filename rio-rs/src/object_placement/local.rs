//! In-memory implementation of the trait [ObjectPlacementProvider]

use std::collections::HashMap;
use std::sync::{Arc, RwLock};

use async_trait::async_trait;

use crate::object_placement::{ObjectPlacement, ObjectPlacementProvider};
use crate::ObjectId;

type PlacementMap = Arc<RwLock<HashMap<String, String>>>;

/// In-memory implementation of the trait [ObjectPlacementProvider]
#[derive(Default, Clone)]
pub struct LocalObjectPlacementProvider {
    placement: PlacementMap,
}

#[async_trait]
impl ObjectPlacementProvider for LocalObjectPlacementProvider {
    async fn update(&self, object_placement: ObjectPlacement) {
        let object_id = format!(
            "{}.{}",
            object_placement.object_id.0, object_placement.object_id.1
        );
        let mut placement_guard = self
            .placement
            .write()
            .expect("Poisoned lock: ObjectPlacementProvider map");
        if let Some(address) = object_placement.server_address {
            *placement_guard.entry(object_id).or_default() = address;
        } else {
            placement_guard.remove(&object_id);
        }
    }

    async fn lookup(&self, object_id: &ObjectId) -> Option<String> {
        let object_id = format!("{}.{}", object_id.0, object_id.1);
        let placement_guard = self
            .placement
            .read()
            .expect("Poisoned lock: ObjectPlacementProvider map");
        placement_guard.get(&object_id).cloned()
    }

    async fn clean_server(&self, address: String) {
        let mut placement_guard = self
            .placement
            .write()
            .expect("Poisoned lock: ObjectPlacementProvider map");
        placement_guard.retain(|_, v| *v != address);
    }

    async fn remove(&self, object_id: &ObjectId) {
        let object_id = format!("{}.{}", object_id.0, object_id.1);
        let mut placement_guard = self
            .placement
            .write()
            .expect("Poisoned lock: ObjectPlacementProvider map");
        placement_guard.remove(&object_id);
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[tokio::test]
    async fn local_object_placement_provider_is_clonable() {
        let provider = LocalObjectPlacementProvider::default();
        let cloned_provider = provider.clone();

        provider
            .update(ObjectPlacement::new(
                ObjectId("test".to_string(), "1".to_string()),
                Some("0.0.0.0:80".to_string()),
            ))
            .await;

        assert!(provider
            .lookup(&ObjectId("test".to_string(), "1".to_string()))
            .await
            .is_some());
        assert!(cloned_provider
            .lookup(&ObjectId("test".to_string(), "1".to_string()))
            .await
            .is_some());

        cloned_provider.clean_server("0.0.0.0:80".to_string()).await;

        assert!(provider
            .lookup(&ObjectId("test".to_string(), "1".to_string()))
            .await
            .is_none());
        assert!(cloned_provider
            .lookup(&ObjectId("test".to_string(), "1".to_string()))
            .await
            .is_none());
    }
}
