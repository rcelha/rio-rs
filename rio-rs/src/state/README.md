Module that provides object persistence

The `rio_rs::state` module has all the traits and structs one needs
to start saving objects from your cluster to persistent storage.

# State

Trait for implementing how to deserialize data in, and serialize data out of
a service (trait object).

Although you need to implement this trait for each state you have in your application,
the `ManagedState` derive implements it for you.

# State Loader

Any type that implements `rio_rs::state_provider::StateLoader`.

The `StateLoader` defines an interface to load state data from a backend.

# State Saver

Any type that implements `rio_rs::state_provider::StateSaver`.

It defines an interface to save data into a persistence backend.

# State Provider

Any type that is both a `State Saver` and a `State Loader`.

It provides functions to load and deserialize state, and serialize and save state.
