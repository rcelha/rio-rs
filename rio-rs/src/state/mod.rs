#![doc = include_str!("README.md")]

use std::ops::Deref;

use crate::errors::LoadStateError;
use crate::registry::IdentifiableType;
use crate::{ServiceObject, WithId};
use async_trait::async_trait;
use serde::de::DeserializeOwned;
use serde::Serialize;

#[cfg(feature = "local")]
pub mod local;

#[cfg(feature = "redis")]
pub mod redis;

// #[cfg(feature = "sql")]
// pub mod sql;

#[cfg(feature = "sqlite")]
pub mod sqlite;

#[cfg(feature = "postgres")]
pub mod postgres;

/// The `StateLoader` defines an interface to load serialized state from a source
///
/// **important** This trait is not responsible for serializing it back to its
/// original type
///
/// TODO use a reader type instead of String on load fn
#[async_trait]
pub trait StateLoader: Send + Sync {
    /// <div class="warning">
    /// TODO
    ///
    /// This here can't be used right now as it depends on the type `T` when
    /// called.
    ///
    /// This makes its usage quite cluncky (if not impossible) as you need to invoke it
    /// with a target of this loader
    /// </div>
    async fn prepare(&self) {}

    async fn load<T>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError>
    where
        T: DeserializeOwned;
}

/// Auto implement [StateLoader] for every type that derefs to a [StateLoader]
///
/// This way you can create a wrapper for a [StateLoader] and it will automatically
/// get this implementation
#[async_trait]
impl<T, S> StateLoader for T
where
    T: Deref<Target = S> + Send + Sync,
    S: StateLoader,
{
    async fn load<O: DeserializeOwned>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<O, LoadStateError> {
        self.deref().load(object_kind, object_id, state_type).await
    }
}

/// The `StateSave` defines an interface to save serialized data into a persistence
/// backend (memory, sql server, etc)
///
/// **important** This trait is not responsible for serializing the state from
/// its original type
#[async_trait]
pub trait StateSaver: Sync + Send {
    async fn prepare(&self) {}

    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError>;
}

/// Auto implement [StateSaver] for every type that derefs to a StateSaver
///
/// This way you can create a wrapper for a [StateSaver] and it will automatically
/// get this implementation
#[async_trait]
impl<T, S> StateSaver for T
where
    T: Deref<Target = S> + Send + Sync,
    S: StateSaver,
{
    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError> {
        self.deref()
            .save(object_kind, object_id, state_type, data)
            .await
    }
}

/// Reponsible for managing states for a specific object
///
/// With this trait one can load/save individual states from an orig (Self) object
#[async_trait]
pub trait ObjectStateManager {
    async fn load_state<T, S>(&mut self, state_loader: &S) -> Result<(), LoadStateError>
    where
        T: IdentifiableType + Serialize + DeserializeOwned + Default, // neends default cause of trait State
        S: StateLoader + Send + Sync,
        Self: State<T> + IdentifiableType + WithId + Send + Sync,
    {
        let object_kind = Self::user_defined_type_id();
        let object_id = self.id();
        let state_type = T::user_defined_type_id();
        let data: T = self
            .load(state_loader, object_kind, object_id, state_type)
            .await
            .or(Err(LoadStateError::ObjectNotFound))?;

        self.set_state(data);
        Ok(())
    }

    async fn save_state<T, S>(&self, state_saver: &S) -> Result<(), LoadStateError>
    where
        T: IdentifiableType + Serialize + DeserializeOwned + Sync + Default, // Needs default cause of trait State
        S: StateSaver,
        Self: State<T> + IdentifiableType + WithId + Send + Sync,
    {
        let object_kind = Self::user_defined_type_id();
        let object_id = self.id();

        let state_type = T::user_defined_type_id();
        let state_value: &T = self.get_state();
        state_saver
            .save(object_kind, object_id, state_type, &state_value)
            .await?;
        Ok(())
    }
}

// If an struct implements ServiceObject, it gets ObjectStateManager out of the box
impl<T> ObjectStateManager for T where T: ServiceObject {}

/// Trait to define how to get and set states in and out of an object
///
/// One need to implement this trait for each state a object holds
#[async_trait]
pub trait State<T>
where
    T: Serialize + DeserializeOwned + Default,
{
    fn get_state(&self) -> &T;
    fn set_state(&mut self, value: T);

    async fn load<S: StateLoader + Sync + Send>(
        &self,
        state_loader: &S,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        state_loader.load(object_kind, object_id, state_type).await
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use rio_macros::{ManagedState, TypeName, WithId};
    use serde::Deserialize;

    type TestResult = Result<(), Box<dyn std::error::Error>>;

    #[derive(Default, Debug, PartialEq, Serialize, Deserialize)]
    struct PersonState {
        name: String,
        age: u8,
    }

    impl IdentifiableType for Option<PersonState> {
        fn instance_type_id(&self) -> &'static str {
            "OptionPersonState"
        }
    }

    #[derive(Default, Debug, Serialize, Deserialize, PartialEq, TypeName)]
    #[rio_path = "crate"]
    struct LegalPersonState {
        legal_name: String,
        id_document: String,
    }

    #[tokio::test]
    async fn sanity_check() -> TestResult {
        let local_state = local::LocalState::new();
        let state = PersonState {
            name: "Foo".to_string(),
            age: 21,
        };
        local_state.save("a", "1", "PersonState", &state).await?;
        let new_state: PersonState = local_state.load("a", "1", "PersonState").await?;
        assert_eq!(state, new_state);
        Ok(())
    }

    #[tokio::test]
    async fn model_call() -> TestResult {
        #[derive(Debug, Default, WithId, TypeName, ManagedState)]
        #[rio_path = "crate"]
        struct Person {
            id: String,
            #[managed_state]
            person_state: Option<PersonState>,
            #[managed_state]
            legal_state: LegalPersonState,
        }
        impl ObjectStateManager for Person {}

        impl Person {
            async fn load_all_states(
                &mut self,
                state_loader: &local::LocalState,
            ) -> Result<(), LoadStateError> {
                let _ = self
                    .load_state::<Option<PersonState>, _>(state_loader)
                    .await?;
                let _ = self.load_state::<LegalPersonState, _>(state_loader).await?;
                Ok(())
            }

            async fn save_all_states<S: StateSaver>(
                &mut self,
                state_saver: &S,
            ) -> Result<(), LoadStateError> {
                self.save_state::<Option<PersonState>, _>(state_saver)
                    .await?;
                self.save_state::<LegalPersonState, _>(state_saver).await?;
                Ok(())
            }
        }

        let local_state = local::LocalState::new();

        {
            let mut person = Person::default();
            person.person_state = Some(PersonState {
                name: "Foo".to_string(),
                age: 22,
            });
            person.legal_state = LegalPersonState {
                legal_name: "Foo Bla".to_string(),
                id_document: "123.123.123-12".to_string(),
            };
            person.save_all_states(&local_state).await?;
        }
        {
            let mut person = Person::default();
            person.load_all_states(&local_state).await?;
            assert!(person.person_state.is_some());
            assert_eq!(&person.legal_state.legal_name, "Foo Bla");
        }
        Ok(())
    }
}
