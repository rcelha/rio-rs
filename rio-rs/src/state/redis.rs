use async_trait::async_trait;
use bb8_redis::{bb8::Pool, RedisConnectionManager};
use redis::AsyncCommands;
use serde::de::DeserializeOwned;
use serde::Serialize;

use super::{StateLoader, StateSaver};
use crate::errors::LoadStateError;

/// State Storage using Redis/Valkey
#[derive(Clone)]
pub struct RedisState {
    pool: Pool<RedisConnectionManager>,
    key_prefix: String,
}

impl RedisState {
    pub async fn from_connect_string(connection_string: &str, key_prefix: Option<String>) -> Self {
        let conn_manager = RedisConnectionManager::new(connection_string).expect("TODO");
        let pool = Pool::builder().build(conn_manager).await.expect("TODO");
        let key_prefix = key_prefix.unwrap_or_default();
        Self { pool, key_prefix }
    }
}

#[async_trait]
impl<T: DeserializeOwned> StateLoader<T> for RedisState {
    async fn load(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let key = format!(
            "{}state:{}:{}:{}",
            self.key_prefix, object_kind, object_id, state_type
        );
        let mut client = self.pool.get().await.map_err(|_| LoadStateError::Unknown)?;
        let se_data: Option<String> = client.get(key).await.expect("TODO");
        if let Some(x) = se_data {
            let data = serde_json::from_str(&x);
            data.map_err(|_| LoadStateError::DeserializationError)
        } else {
            Err(LoadStateError::ObjectNotFound)
        }
    }
}

#[async_trait]
impl<T: Serialize + Send + Sync> StateSaver<T> for RedisState {
    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &T,
    ) -> Result<(), LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let key = format!(
            "{}state:{}:{}:{}",
            self.key_prefix, object_kind, object_id, state_type
        );
        let ser_data =
            serde_json::to_string(&data).map_err(|_| LoadStateError::SerializationError)?;
        let mut client = self.pool.get().await.map_err(|_| LoadStateError::Unknown)?;
        let _: () = client
            .set(key, ser_data)
            .await
            .map_err(|_| LoadStateError::Unknown)?;
        Ok(())
    }
}
