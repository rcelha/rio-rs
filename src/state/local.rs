use super::{StateLoader, StateSaver};
use crate::errors::LoadStateError;
use async_trait::async_trait;
use dashmap::DashMap;
use serde::de::DeserializeOwned;
use serde::Serialize;

/// `LocalState` is a state provider for testing purposes
///
/// It stores all the serialized states into a single `DashMap`
#[derive(Debug)]
pub struct LocalState {
    data: DashMap<(String, String, String), String>,
}

impl LocalState {
    pub fn new() -> LocalState {
        LocalState {
            data: DashMap::new(),
        }
    }
}

#[async_trait]
impl StateLoader for LocalState {
    async fn load<T: DeserializeOwned>(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
    ) -> Result<T, LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let k = (object_kind, object_id, state_type);

        if let Some(x) = self.data.get(&k) {
            Ok(serde_json::from_str(&x).expect("TODO"))
        } else {
            Err(LoadStateError::Unknown)
        }
    }
}

#[async_trait]
impl StateSaver for LocalState {
    async fn save(
        &self,
        object_kind: &str,
        object_id: &str,
        state_type: &str,
        data: &(impl Serialize + Send + Sync),
    ) -> Result<(), LoadStateError> {
        let object_kind = object_kind.to_string();
        let object_id = object_id.to_string();
        let state_type = state_type.to_string();
        let k = (object_kind, object_id, state_type);
        self.data
            .insert(k, serde_json::to_string(&data).expect("TODO"));
        Ok(())
    }
}

#[cfg(test)]
mod test {
    use rio_macros::TypeName;
    use serde::Deserialize;

    use super::*;

    type TestResult = Result<(), Box<dyn std::error::Error>>;

    #[derive(TypeName, Debug, Serialize, Deserialize, PartialEq)]
    #[rio_path = "crate"]
    struct TestState {
        name: String,
    }

    #[tokio::test]
    async fn sanity_check() -> TestResult {
        let local_state = LocalState::new();
        let state = TestState {
            name: "Foo".to_string(),
        };
        local_state.save("a", "1", "TestState", &state).await?;
        let new_state: TestState = local_state.load("a", "1", "TestState").await?;
        assert_eq!(state, new_state);
        Ok(())
    }
}
