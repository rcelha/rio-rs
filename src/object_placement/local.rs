use async_trait::async_trait;

use crate::object_placement::{ObjectPlacement, ObjectPlacementProvider};
use crate::ObjectId;

#[derive(Default)]
pub struct LocalObjectPlacementProvider {}

#[async_trait]
impl ObjectPlacementProvider for LocalObjectPlacementProvider {
    async fn update(&self, _object_placement: ObjectPlacement) {}
    async fn lookup(&self, _object_id: &ObjectId) -> Option<String> {
        None
    }
    async fn clean_silo(&self, _address: String) {}
    async fn remove(&self, _object_id: &ObjectId) {}
}
