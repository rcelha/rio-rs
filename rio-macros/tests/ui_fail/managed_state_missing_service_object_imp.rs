use rio_macros::*;
use rio_rs::state::local::LocalState;
use rio_rs::state::ObjectStateManager;

#[derive(TypeName, serde::Serialize, serde::Deserialize, Default)]
struct TestVec(Vec<usize>);

#[derive(Default, WithId, TypeName, ManagedState)]
struct TestService {
    id: String,
    #[managed_state(provider = LocalState)]
    tests: TestVec,
}

// This is the missing bit:
// impl ServiceObject for TestService {}

fn main() {}
