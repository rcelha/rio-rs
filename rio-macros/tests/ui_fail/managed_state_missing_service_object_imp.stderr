error[E0599]: the method `load_state` exists for mutable reference `&mut TestService`, but its trait bounds were not satisfied
 --> tests/ui_fail/managed_state_missing_service_object_imp.rs:8:37
  |
8 | #[derive(Default, WithId, TypeName, ManagedState)]
  |                                     ^^^^^^^^^^^^ method cannot be called on `&mut TestService` due to unsatisfied trait bounds
9 | struct TestService {
  | ------------------ doesn't satisfy `TestService: ObjectStateManager` or `TestService: ServiceObject`
  |
  = note: the following trait bounds were not satisfied:
          `TestService: ServiceObject`
          which is required by `TestService: ObjectStateManager`
          `&mut TestService: ServiceObject`
          which is required by `&mut TestService: ObjectStateManager`
note: the trait `ServiceObject` must be implemented
 --> $WORKSPACE/rio-rs/src/service_object.rs
  |
  | pub trait ServiceObject: Default + WithId + IdentifiableType {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `load_state`, perhaps you need to implement it:
          candidate #1: `ObjectStateManager`
  = note: this error originates in the derive macro `ManagedState` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused import: `rio_rs::state::ObjectStateManager`
 --> tests/ui_fail/managed_state_missing_service_object_imp.rs:3:5
  |
3 | use rio_rs::state::ObjectStateManager;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
